Implement Metadata to emit runtime extra

uranusjr opened the pull request:
Close #37810. This implements the `yield` syntax and the `Metadata` class, and mechanism to collect the yielded values in the worker.

uranusjr committed:
Implement Metadata to emit runtime extra

The dataset URI coercing logic has been extracted into its own
(internal) function for reuse.

uranusjr committed:
Collect yielded metadata during task execution

uranusjr committed:
Allow task hooks to be generator functions

uranusjr committed:
Detect generator function in PythonOperator

Also reverted check for task.pre_execute and task.post_execute since
those are defined on BaseOperator and shouldn't be a generator function.
Instead, those functions need to individually check if the hooks wrapped
inside are generator functions.

uranusjr committed:
Document 'yield Metadata(...)'

uranusjr commented on airflow/operators/python.py:
```diff
@@ -243,13 +243,14 @@ def execute(self, context: Context) -> Any:
         return KeywordParameters.determine(self.python_callable, self.op_args, context).unpacking()
 
-    def execute_callable(self) -> Any:
+    def execute_callable(self, dataset_events: DatasetEventAccessors) -> Any:
```
I had to change the signature here (and in subclasses) so the callable can access the accessors. I _think_ this shouldn’t be public and thus not backward incompatible? Not exactly sure…

jscheffl replied:
Mhm, it is not exactly defined as "public" but at least we have also a custom subclass of python operator developed in our environment. So everybody having a subclass of this will see a breaking change?

eladkal commented on docs/apache-airflow/authoring-and-scheduling/datasets.rst:
```diff
@@ -232,20 +232,36 @@ Attaching extra information to an emitting Dataset Event
+
+Airflow automatically collects all yielded metadata, and populates dataset events with extra information for corresponding metadata objects.
+
+This can also be done in classic operators. The best way is to subclass the operator and override ``execute``. Alternatively, extras can also be added in a task's ``pre_execute`` or ``post_execute`` hook.
```
pre and post don't have retry mechanism so I assume there can be cases where not all data collected. So there are pros and cons if users decided to do it in execute() or in pre/post

uranusjr replied:
I added a note about this.

eladkal reviewed (approved):
Nice!

uranusjr committed:
Handle when context is missing dataset_events

The key is not always available. If it isn't, we just create an accessor
object on the fly.

uranusjr committed:
Add a note about execution hooks' rerun issue

jscheffl commented on airflow/operators/python.py:
```diff
@@ -232,7 +232,11 @@ def execute(self, context: Context) -> Any:
+            dataset_events = context["dataset_events"]
+        except KeyError:
+            dataset_events = context["dataset_events"] = DatasetEventAccessors()
+        return_value = self.execute_callable(dataset_events)
```
I am not sure why we need that change to add `dataset_events` to the calling args - we don't add other context information here as well. Why should be dataset event be different than other context vars?

uranusjr replied:
Because `execute_callable` does not receive the context at all. If you ask me it’s sort of designed quite weirdly to begin with.

I think I have a way to avoid not passing in the additional argument, by storing the value on the operator (`self`) instead. It’s a bit hacky and more of a workaround.

jscheffl reviewed (commented):
I understand the desire to add the option to `yield` but I do not get the relation why the execution internals of PythonOperator implementation needs a change in method signatures for this. I feel that DataSetEvents should be the same approach like other context information and do not see why the callable execution needs a signature change.

uranusjr committed:
Store dataset_events on the operator instead

uranusjr committed:
Fix lambda exception test

uranusjr committed:
Add dataset_events to PVO test

uranusjr committed:
Do not insert the accessors back into context

uranusjr committed:
Make dataset_events access safer

The key is not always available (mostly in tests). Create a placeholder
if it's not.

jscheffl reviewed (approved):
Thanks for the rework. Now I really "like" it. Before I would have stepped back from review but would have not blocked merge :-D

One minor thing would be great to also try to use it in one of the example DAGs for example in example_producer_1 or 2? Does it make sense to add it one of the examples? (I propose this on top of docs which are good, sometimes people seek for inspiration or examples in the example code. Not many people read docs before designing or coding.
